src/exampleWorker.js

```diff
@@ -0,0 +1,229 @@
+      }
+    }
+  };
+
+  self.addEventListener('message', function(e) {
+    postMessage(e.data.name);
+    switch (e.data.name) {
+      case 'initialize':
+        /* eslint-disable no-eval */
+        grammar = ohm.makeRecipe(eval(e.data.recipe));
+        /* eslint-enable no-eval */
+
+        start();
+        break;
+      case 'request:examples':
+        var ruleName = e.data.args[0];
+
+        var examplesForRule = examplePieces[ruleName] || null;
+        self.postMessage({name: 'received:examples',
+                          args: [ruleName, examplesForRule]});
+        break;
+      case 'update:neededExamples':
+        var neededExamples = utils.difference(
+          Object.keys(grammar.rules),
+          Object.keys(examplePieces)
+        );
+
+        self.postMessage({name: 'received:neededExamples',
+                          args: [neededExamples]});
+        break;
+      case 'add:userExample':
+        ruleName = e.data.args[0];
+        var example = e.data.args[1];
+        generator.processExampleFromUser(example, ruleName);
+        break;
+    }
+  }, false);
+
+  function addPiecesToDict(trace, examples) {
+    if (trace.expr.constructor.name === 'Terminal') {
+      return;
+    } else {
+      if (trace.expr.constructor.name === 'Apply') {
+        var ruleName = trace.expr.toString();
+        if (!examples.hasOwnProperty(ruleName)) {
+          examples[ruleName] = [];
+        }
+
+        if (!examples[ruleName].includes(trace.source.contents)) {
+          examples[ruleName].push(trace.source.contents);
+        }
+      }
+      trace.children
+           .filter(function(child) { return child.succeeded; })
+           .forEach(function(child) { return addPiecesToDict(child, examples); });
+    }
+  }
+
+  function ExampleGenerator(examplePieces) {
+    this.examplePieces = examplePieces;
+    this.rules = initialRules(grammar);
+    this.examplesNeeded = utils.difference(
+      Object.keys(grammar.rules),
+      Object.keys(examplePieces)
+    );
+    this.currentRuleIndex = 0;
+  }
+
+  ExampleGenerator.prototype.processExampleFromUser = function(example, optRuleName) {
+    if (optRuleName) {
+      var trace = grammar.trace(example, optRuleName);
+      if (trace.succeeded) {
+        addPiecesToDict(trace, examplePieces);
+      }
+    }
+
+    utils.objectForEach(grammar.rules, function(ruleName) {
+      var trace = grammar.trace(example, ruleName);
+      if (trace.succeeded) {
+        addPiecesToDict(trace, examplePieces);
+      }
+    });
+
+    var oldSize = this.examplesNeeded.length;
+    this.examplesNeeded = this.examplesNeeded.filter(function(ruleName) {
+      return !examplePieces.hasOwnProperty(ruleName);
+    });
+    if (this.examplesNeeded.length < oldSize) {
+      self.postMessage({name: 'received:neededExamples',
+                        args: [this.examplesNeeded]});
+    }
+  };
+
+  ExampleGenerator.prototype.next = function() {
+    var that = this;
```

Comment:
So, after a bit of thought, in this case, I think I would prefer to use repeat instead of for, just because it's more declarative. There's something similar to repeat strings in common.js. Would that be ok?